### DISCUSSION TITLE
How run commands (pm install) when adb shell need enter password

### DISCUSSION THREAD

**Discussion by dstarkoff** [yume-chan/ya-webadb#678] [Q&A] [answered]
After push apk to tmp folder, **pm** will run `shell pm install *fileName.apk*` command, but my device needs enter password for shell...
in code im try `pm.pushAndInstallStream(stream)`

**Comment by yume-chan** [comment]
Can you capture a video of how you install apks using ADB in a terminal? Preferable with a separate keyboard input overlay if the password is not echoed.

Also, what's the model of your device?

**Reply by dstarkoff** [reply]
Im install on Changan car head unit (Huawei board with MTK cpu). Not root, not unlock bootloader.

I will capture video later. Step-by-step process on terminal:

`adb devices` // check the device is connected
`adb push fileName.apk /data/local/tmp` // push file to device file system
`adb shell pm install /data/local/tmp/fileName.apk` // run install from shell
`please input verify password:` // enter password **adb36987** and press Enter
`verify success!` // this app start install on device

password permanently **adb36987**, need enter when i use **shell**

**Reply by dstarkoff** [reply]
https://github.com/user-attachments/assets/82f53de6-d2d9-44e5-a703-864a3ae6b1cc

**Comment by dstarkoff** [comment, marked as answer]
Im have a solution (use core functions of package):
```
const process = await adb.subprocess.shell('pm install -r ' + `/data/local/tmp/${file.name}`),
          writer = process.stdin.getWriter(),
          decoder = new TextDecoder();
    let   accumulatedOutput = '';

    setInstalStatus("installing");

    process.stdout.pipeTo(
      new WritableStream({
        async write(chunk) {
          const output = decoder.decode(chunk);
          accumulatedOutput += output; // console.log(accumulatedOutput)
          if (accumulatedOutput.includes("please input verify password:")) {
            await writer.write(new TextEncoder().encode("adb36987\n"));
            accumulatedOutput = '';
          }
          if (accumulatedOutput.includes("verify success!")) {
            accumulatedOutput = '';
          }
          if (accumulatedOutput.includes("Success")) {
            accumulatedOutput = '';
            console.log("Установлено!")
            setInstalStatus("default")
            await writer.close();
          }
        },
      })
    );
  };
```
